File: Convolutional_Ex/Handsome_Tris_BigEx.cpp
#include <iostream>
#include <string.h>
#include <iomanip>

using namespace std;

int main(){
    int m,n,p,q,sum=0;
    cout << "input m: ";
    cin >> m;
    cout << "input n: ";
    cin >> n;
    cout << "input p: ";
    cin >> p;
    cout << "input q: ";
    cin >> q;
    int keras[p][q];
    int source[m][n];
    if (m<p || n<q){
        cout << "You dumb AF, GET OUT !" << endl;
    }
    else{
        cout << "Input source !" << endl;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++){
                cout << "Row " << i+1 << ", Column " << j+1 << " : "; 
                cin >> source[i][j];
            }
        }
        cout << "Input kernel !" << endl;
        for (int i = 0; i < p; i++){
            for (int j = 0; j < q; j++){
                cout << "Row " << i+1 << ", Column " << j+1 << " : "; 
                cin >> keras[i][j];
            }
        }

        if (m==p && n==q){
            for (int i = 0; i < p; i++){
                for (int j=0; j < q; j++){
                    sum += source[i][j] * keras[i][j];
                }
            }
            cout << "Output: " << sum << endl;
        }
        else {
            int result[m-p+1][n-q+1];
            int cor=0, sair=0;
            while (1){
                sum = 0;
                for (int i = 0; i < p; i++){
                    for (int j=0; j < q; j++){
                        sum += source[i + cor][j + sair] * keras[i][j];
                    }
                }
                result[cor][sair] = sum;
                if (p + sair == m){
                    sair = 0;
                    if (q + cor == n){
                        cout << "Thanh !!!!!!!!!!!!" << endl;
                        break;
                    }
                    cor += 1;
                }
                else{
                    sair++;
                }
            }
            for (int i = 0; i < m-p+1; i++){
                for (int j = 0; j < n-q+1; j++){
                    cout << setw(4) << result[i][j] << " ";
                }
                cout << endl;
            }
        }
    }
    return 0;
}

File: Convolutional_Ex/Handsome_Tris_BigEx.py
def mini_convolution(p,q,sair,cor):
    sum = 0
    for row in range(p):
        for col in range(q):
            sum += source[row + cor][col + sair] * keras[row][col]
    return sum


m = int(input('input m:'))
n = int(input('input n:'))
p = int(input('input p:'))
q = int(input('input q:'))

if m<p or n<q:
    print('You dumb AF, GET OUT !')     
else:
    source = [[0 for i in range(m)] for i in range(n)]
    keras = [[0 for i in range(p)] for i in range(q)]
    print('Nhập vào source:')
    for row in range(m):
        for col in range(n):
            i = input(f'Nhập vào giá trị tại hàng {row+1} và cột {col+1}:')
            source[row][col] = int(i)
    print('Nhập vào keras:')
    for row in range(p):
        for col in range(q):
            i = input(f'Nhập vào giá trị tại hàng {row+1} và cột {col+1}:')
            keras[row][col] = int(i)
                    
    if m==p and n==q:
        sum = mini_convolution(p,q,0,0)
        print('Kết quả =',sum)
    else:
        result = [[0 for i in range(m-p+1)] for i in range(n-q+1)]
        cor = sair = 0
        while True:
            sum = mini_convolution(p,q,sair,cor)
            result[cor][sair] = int(sum)
            if p + sair == m:
                sair = 0
                if q + cor == n:
                    print('Trí siêu đẹp trai !!!!!!!!!!!!!!!')
                    for row in result:
                        print(row)
                    break
                cor += 1
            else:
                sair += 1
        
    
    